internal/app/machined/pkg/controllers/config/node_labels_config.go

```diff
@@ -0,0 +1,78 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package config
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/cosi-project/runtime/pkg/controller"
+	"github.com/cosi-project/runtime/pkg/resource"
+	"github.com/cosi-project/runtime/pkg/state"
+	"github.com/siderolabs/go-pointer"
+	"go.uber.org/zap"
+
+	"github.com/talos-systems/talos/pkg/machinery/resources/config"
+)
+
+// NodeLabelsConfigController manages config.NodeLabelsConfig based on configuration.
+type NodeLabelsConfigController struct{}
+
+// Name implements controller.Controller interface.
+func (ctrl *NodeLabelsConfigController) Name() string {
+	return "config.NodeLabelsConfigController"
+}
+
+// Inputs implements controller.Controller interface.
+func (ctrl *NodeLabelsConfigController) Inputs() []controller.Input {
+	return []controller.Input{
+		{
+			Namespace: config.NamespaceName,
+			Type:      config.MachineConfigType,
+			ID:        pointer.To(config.V1Alpha1ID),
+			Kind:      controller.InputWeak,
+		},
+	}
+}
+
+// Outputs implements controller.Controller interface.
+func (ctrl *NodeLabelsConfigController) Outputs() []controller.Output {
+	return []controller.Output{
+		{
+			Type: config.NodeLabelsConfigType,
```

Comment:
it's not consistent in the codebase right now, but looks like most of other `k8s.` configuration resources are in the `k8s/` package